---
title: "class 13"
author: "Miso Lee"
date: "May 15, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 1000 Genomes Data Analysis

Downloaded genotype data from Ensembe for Asthma associated SNP rs8067378
See: (MXL) Mexican Ancestry in Los Angeles, California

```{r}
# Read CSV file
genotype <- read.csv("373531-SampleGenotypes-Homo_sapiens_Variation_Sample_rs8067378.csv")
head(genotype)
```

Find how many of each genotype there are in this dataset

```{r}
genotype.summary <- table(genotype[,2])
genotype.summary
```

Proportion of each genotype
```{r}
# What proportion of the Mexican Ancestry in Los Angeles sample population (MXL) are homozygous for the asthma associated SNP (G|G)? 14%
genotype.summary/nrow(genotype)
```

## RNA-Seq

Look more closely at HG00109 with G|G genotype for this SNP (from GBR population).

## Side-note on Fastq quality scores
```{r}
# install.packages("seqinr")
# install.packages("gtools")
library(seqinr)
library(gtools)
phred <- asc( s2c("DDDDCDEDCDDDDBBDDDCC@") ) - 33
phred
```

```{r}
prob <- 10**(-phred/10)
prob
```

## Section 4. RNA-Seq Population Scale Analysis

Read our results file from RNA-seq  across the ~230 samples (i.e. people!)

```{r}
# read file into R
url <- "https://bioboot.github.io/bimm143_S18/class-material/rs8067378_ENSG00000172057.6.txt"
expr <- read.table(url)
```

Look at data
```{r}
table(expr$geno)
```

How many samples?
```{r}
nrow(expr)
```

Expression level summary stats
```{r}
summary(expr)
```

```{r}
summary(expr$exp[(expr$geno == "G/G")])
summary(expr$exp[(expr$geno == "A/G")])
summary(expr$exp[(expr$geno == "A/A")])
```

```{r}
# boxplot
boxplot(exp ~ geno, data=expr, notch=TRUE)
```

```{r}
library(ggplot2)
ggplot(expr, aes(geno, exp)) + geom_boxplot()
```

```{r}
## Histogram of the exp column with ggplot2
ggplot(expr, aes(exp, fill = geno)) + geom_density(alpha = 0.2)
```

Add another cool plot
```{r}
# Boxplot with the data shown
ggplot(expr, aes(geno, exp, fill=geno)) + 
  geom_boxplot(notch=TRUE, outlier.shape = NA) + 
  geom_jitter(shape=16, position=position_jitter(0.2), alpha=0.4)
```

